services/web-app/utils/use-query-state.js

```diff
@@ -57,15 +28,35 @@ export const useQueryState = (
       return null
     }
 
-    const query = new URLSearchParams(window.location.search)
-    const queryValue = query.get(key)
-    const storageValue = saveToStorage ? localStorage.getItem(`${router.pathname}:${key}`) : null
+    const query = queryString.parseUrl(window.location.search, {
+      ...queryStringConfig,
+      parseNumbers,
+      parseBooleans
+    }).query
+
+    const storageValue = saveToStorage
+      ? localStorage.getItem(`${router.pathname}:${key}`)
+      : undefined
+
+    let val = query[key]
 
-    return queryValue !== null ? parse(queryValue) : storageValue
-  }, [key, parse, router.pathname, saveToStorage])
+    if (val !== undefined && validateValue && !validateValue(val)) {
+      val = undefined
+      delete query[key]
 
-  // Update the "state" when the router.query key changes
-  const value = useMemo(getValue, [getValue, router.query[key]])
+      // Change query state without rerendering page
+      history.replaceState(null, null, `?${queryString.stringify(query, queryStringConfig)}`)
+    }
+
+    if (val === undefined) {
+      val = storageValue
+    }
```

Comment:
It looks like this is reverting to the value form local storage if the supplied value was either absent or failed validation. What happens in the case where you're _not_ saving to local storage and the value becomes invalid (after having previously become valid)? Would the application see "undefined" now for the value?

I think this hook would benefit from top-level documentation describing the exact contract it provides, including these types of edge cases.

Reply:
correct, the value would be undefined, that makes sense to me, should we code for a different behavior? 
I'll add some top level comments regarding the hook's functionality and how it behaves 👍

Reply:
Added JSDocs!